exonum_derive/src/lib.rs

```diff
@@ -0,0 +1,302 @@
+            let field = v
+                .fields
+                .iter()
+                .next()
+                .expect("TransactionSet enum variant can't be empty");
+            (n as u16, v.ident.clone(), field.ty.clone())
+        }).collect::<Vec<_>>();
+
+    let convert_1 = variants.iter().map(|(_, id, ty)| {
+        quote! {
+          impl Into<#name> for #ty {
+               fn into(self) -> #name {
+                     #name::#id(self)
+               }
+          }
+
+          impl Into<ServiceTransaction> for #ty {
+              fn into(self) -> ServiceTransaction {
+                  let set: #name = self.into();
+                  set.into()
+              }
+          }
+        }
+    });
+
+    let into_service_tx = {
+        let tx_set_impls = variants.iter().map(|(n, id, _)| {
+            quote! {
+                #name::#id(ref tx) => (#n, tx.encode().unwrap()),
+            }
+        });
+
+        quote! {
+            impl Into<ServiceTransaction> for #name {
+                fn into(self) -> ServiceTransaction {
+                    let (id, vec) = match self {
+                        #( #tx_set_impls )*
+                    };
+                    ServiceTransaction::from_raw_unchecked(id, vec)
+                }
+            }
+        }
+    };
+
+    let tx_set_impl = {
+        let tx_set_impls = variants.iter().map(|(n, id, ty)| {
+            quote! {
+                #n => {
+                    Ok(#name::#id(#ty::decode(&vec)?))
+                },
+            }
+        });
+
+        quote! {
+
+            impl TransactionSet for #name {
+                fn tx_from_raw(raw: RawTransaction) -> Result<Self, encoding::Error> {
+                    let (id, vec) = raw.service_transaction().into_raw_parts();
+                    match id {
+                        #( #tx_set_impls )*
+                        num => Err(encoding::Error::Basic(
+                            format!(
+                                "Tag {} not found for enum {}.",
+                                num, stringify!(#name)
+                            ).into(),
+                        )),
+                    }
+                }
+            }
+
+        }
+    };
+
+    let into_boxed_tx = {
+        let tx_set_impls = variants.iter().map(|(_, id, _)| {
+            quote! {
+                #name::#id(tx) => Box::new(tx),
+            }
+        });
+
+        quote! {
+            impl Into<Box<dyn Transaction>> for #name {
+                fn into(self) -> Box<dyn Transaction> {
+                    match self {
+                        #( #tx_set_impls )*
+                    }
+                }
+            }
+        }
+    };
+
+    let cr = if env::var("CARGO_PKG_NAME").unwrap() == "exonum" {
+        quote!(crate)
+    } else {
+        quote!(exonum)
+    };
+
+    let expanded = quote! {
+        mod #mod_name{
+            use super::*;
+            use #cr::blockchain::{Transaction, TransactionSet};
```

Comment:
I'm semi-against using *any* imports (except for `super::*`), as they may lead to shadowing and/or weird compile errors.